### PR TITLE
 Advanced OOP

### LINKED ISSUE #121: Classes (advanced)
- Inheritance and abstract classes
- Setters and getters, encapsulation in Python
- Write better classes with `dataclasses` and `attrs`
- Useful decorators (`@classmethod`, `@staticmethod`, ...)
- Abstract classes (`ABC`)

### PR DESCRIPTION
**Pull request by despadam**
fixes #121

### CODE REVIEW

File: object_oriented_programming_advanced.ipynb
Comment (edoardob90) [inline object_oriented_programming_advanced.ipynb:1]: I made changes to the notebook (mostly rewordings) in a separate commit. Changes were simple enough that was much quicker.

**Review by edoardob90 — CHANGES_REQUESTED** [review]
Thanks, @despadam 👏🏻  Thorough and the exercises are very much in line with the topics you covered.

I left you a few comments on how to improve the tests.

File: object_oriented_programming_advanced.ipynb
Comment (despadam) [reply object_oriented_programming_advanced.ipynb:1]: Looks good, thanks a lot!

**Comment by despadam** [comment]
Thanks @edoardob90 for your comments! They have all been addressed. Could you please do a last check that everything works properly?

**Review by edoardob90 — APPROVED** [review]
Great! 💯

**Pull request merged**